test/test_parse_array.py

```diff
@@ -119,3 +132,93 @@ def test_transform_layered_array():
     assert result["data"].shape == (2, 8)
     assert result["data"].dims == ("layer", "dim_0")
     assert np.array_equal(result["data"][0], np.ones((8,)))
+
+
+def test_transform_full_component():
+    grammar = """
+start: block*
+block: options_block | arrays_block
+options_block: "begin"i "options"i options_vars "end"i "options"i
+options_vars: (r2d2 | b | c | p)*
+arrays_block: "begin"i "arrays"i arrays_vars "end"i "arrays"i
+arrays_vars: (x | y | z)*
+r2d2: "r2d2"i // keyword
+b: "b"i string
+c: "c"i integer
+p: "p"i double
+x: "x"i array
+y: "y"i array
+z: "z"i array
+"""
+    parser = make_typed_parser(grammar)
+
+    class BlockTransformer(TypedTransformer):
+        def start(self, items: list[Any]) -> dict:
+            return ChainMap(*items)
+
+        def block(self, items: list[Any]) -> dict:
+            return items[0]
+
+        def options_block(self, items: list[Any]) -> dict:
+            return {"options": items[0]}
+
+        def arrays_block(self, items: list[Any]) -> dict:
+            return {"arrays": items[0]}
+
+        def options_vars(self, items: list[Any]) -> dict:
+            return {item[0].lower(): item[1] for item in items}
+
+        def arrays_vars(self, items: list[Any]) -> dict:
+            return {item[0].lower(): item[1] for item in items}
+
+        def r2d2(self, _: list[Any]) -> bool:
```

Comment:
instead of hardcoding these in future, we can inject class definitions into the transformer, and have a single hook which handles blocks, variables, etc as appropriate based on the class definition